File: 算法/二分/找出第 K 小的数对距离/找出第 K 小的数对距离.cpp
// https://leetcode.cn/problems/find-k-th-smallest-pair-distance/

class Solution {
public:
    int smallestDistancePair(vector<int>& nums, int k) {
        sort(nums.begin(), nums.end());
        int L = 0, R = nums.back()-nums.front();
        while (L <= R)
        {
            int M = (L + R) >> 1, cot = 0;
            for (int i=0; i<nums.size(); i++)
            {
                int l = 0, r = i;
                while (l <= r)
                {
                    int m = (l + r) >> 1;
                    if (nums[i]-nums[m] > M) l = m + 1;
                    else r = m - 1;
                }
                cot += i - l;
            }
            if (cot >= k) R = M - 1;
            else L = M + 1;
        }
        return L;
    }
};

File: 算法/二分/找出第 K 小的数对距离/找出第 K 小的数对距离.py
# https://leetcode.cn/problems/find-k-th-smallest-pair-distance/

class Solution:
    def smallestDistancePair(self, nums: List[int], k: int) -> int:
        nums, n = sorted(nums), len(nums)
        L, R = 0, nums[n-1] - nums[0]
        while L <= R:
            M, C = (L + R) // 2, 0
            for i in range(n):
                l, r = 0, i
                while l <= r:
                    m = (l + r) // 2
                    if nums[i]-nums[m] > M: l = m + 1
                    else: r = m - 1
                C += i - l
            if C >= k: R = M - 1
            else: L = M + 1
        return L